handsanitizer/FunctionCallerGenerator.cpp

```diff
@@ -122,95 +122,86 @@ namespace handsanitizer {
-
-        // if scalar
-        if (!arg.isArrayTy() && !arg.isStructTy() && !arg.isPointerTy()) {
-            output << "\"" << arg.getCTypeName() << "\"";
+            return getJsonInputTemplateTextForJsonRvalueAsJson(*arg.getPointerElementType(), typePath, pointerIndirections + 1);
+
+        if (arg.isScalarTy()) {
+            if (pointerIndirections > 0)
+                json["meta_data"]["type"] = arg.getCTypeName() + std::string(pointerIndirections, '*');
+            else
+                json["meta_data"]["type"] = arg.getCTypeName();
+            json["value"] = nullptr;
         }
-
         if (arg.isArrayTy()) {
-            output << "[";
+            if (pointerIndirections > 0)
+                json["meta_data"]["type"] = arg.getTypeName() + std::string(pointerIndirections, '*');
+            else
+                json["meta_data"]["type"] = arg.getTypeName();
+            json["meta_data"]["element_type"] = arg.getArrayElementType()->getCTypeName();
+            json["meta_data"]["size"] = arg.getArrayNumElements();
+
+            nlohmann::json arrayJson;
             for (int i = 0; i < arg.getArrayNumElements(); i++) {
-                output << getJsonInputTemplateTextForJsonRvalue(*arg.getArrayElementType(), typePath);
-                if (i != arg.getArrayNumElements() - 1) {
-                    output << ",";
-                }
+                arrayJson.emplace_back(getJsonInputTemplateTextForJsonRvalueAsJson(*arg.getArrayElementType(), typePath));
             }
-            output << "]";
+            json["value"] = arrayJson;
         }
 
         if (arg.isStructTy()) {
-            output << "{";
+            if (pointerIndirections > 0)
+                json["meta_data"]["type"] = arg.getTypeName() + std::string(pointerIndirections, '*');
+            else
+                json["meta_data"]["type"] = arg.getTypeName();
 
+            json["meta_data"]["struct_name"] = arg.getCTypeName();
 
             for (auto &mem : arg.getNamedMembers()) {
                 std::vector<std::pair<Type *, std::string>>::iterator pathToPreviousUnrollingOfType;
                 pathToPreviousUnrollingOfType = std::find_if(typePath.begin(), typePath.end(),
-                                                             [mem](const std::pair<Type *, std::string>& pair) { return mem.type == pair.first; });
-                if(pathToPreviousUnrollingOfType != typePath.end()){
+                                                             [mem](const std::pair<Type *, std::string> &pair) { return mem.type == pair.first; });
+                if (pathToPreviousUnrollingOfType != typePath.end()) {
                     // member type is already found
                     std::stringstream pathToFirstOccurrence;
-                    for(auto& path : typePath)
+                    for (auto &path : typePath)
                         pathToFirstOccurrence << "[\'" << path.second << "\']";
```

Comment:
Could this also be built using the nlohmann APIs?

Reply:
Yes and no. The value here is used as a reference to users as to which path they have to walk from the input template to get to the first occurrence of the type to which it is self referential with. 

As we have an option to specify meta data now, I believe it would actually better to change this such that each node in the json tree has an numeric id and refer back to it with that. As that should be much easier to parse.